src/filters/capture.rs

```diff
@@ -14,78 +14,80 @@
+use self::{
+    affix::{Prefix, Suffix},
+    metrics::Metrics,
+    regex::Regex,
+};
 
+use self::quilkin::extensions::filters::capture::v1alpha1 as proto;
 pub use config::{Config, Strategy};
 
-pub const NAME: &str = "quilkin.extensions.filters.capture_bytes.v1alpha1.CaptureBytes";
+pub const NAME: &str = "quilkin.extensions.filters.capture.v1alpha1.Capture";
 
 /// Creates a new factory for generating capture filters.
 pub fn factory() -> DynFilterFactory {
     Box::from(CaptureBytesFactory::new())
 }
 
-struct CaptureBytes {
-    capture: Box<dyn Capture + Sync + Send>,
+/// Trait to implement different strategies for capturing packet data.
+pub trait CaptureStrategy {
+    /// Capture packet data from the contents, and optionally returns a value if
+    /// anything was captured.
+    fn capture(&self, contents: &mut Vec<u8>, metrics: &Metrics) -> Option<Value>;
+}
+
+struct Capture {
+    capture: Box<dyn CaptureStrategy + Sync + Send>,
     /// metrics reporter for this filter.
     metrics: Metrics,
     metadata_key: Arc<String>,
-    size: usize,
-    remove: bool,
+    is_present_key: Arc<String>,
 }
 
-impl CaptureBytes {
+impl Capture {
     fn new(config: Config, metrics: Metrics) -> Self {
-        CaptureBytes {
-            capture: config.strategy.as_capture(),
+        Self {
+            capture: config.strategy.into_capture(),
             metrics,
+            is_present_key: Arc::new(config.metadata_key.clone() + "/is_present"),
             metadata_key: Arc::new(config.metadata_key),
-            size: config.size,
-            remove: config.remove,
         }
     }
 }
 
-impl Filter for CaptureBytes {
+impl Filter for Capture {
     #[cfg_attr(feature = "instrument", tracing::instrument(skip(self, ctx)))]
     fn read(&self, mut ctx: ReadContext) -> Option<ReadResponse> {
-        // if the capture size is bigger than the packet size, then we drop the packet,
-        // and occasionally warn
-        if ctx.contents.len() < self.size {
-            if self.metrics.packets_dropped_total.get() % 1000 == 0 {
-                warn!(count = ?self.metrics.packets_dropped_total.get(), "Packets are being dropped due to their length being less than {} bytes", self.size);
-            }
-            self.metrics.packets_dropped_total.inc();
-            return None;
-        }
-        let token = self
-            .capture
-            .capture(&mut ctx.contents, self.size, self.remove);
-
+        let capture = self.capture.capture(&mut ctx.contents, &self.metrics);
         ctx.metadata
-            .insert(self.metadata_key.clone(), Value::Bytes(token.into()));
+            .insert(self.is_present_key.clone(), Value::Bool(capture.is_some()));
 
-        Some(ctx.into())
+        if let Some(value) = capture {
+            ctx.metadata.insert(self.metadata_key.clone(), value);
+            Some(ctx.into())
+        } else {
+            None
```

Comment:
I just noticed this change in behaviour - this would mean that if there is no match, we drop the whole packet.

I don't think we want to change the behaiviour here. The job of this filter is purely to capture data - it's not to make decisions about it.

🤔 Unless we want to add an option to "drop on failing to match" - which I think would be a nice addition (we could do this in a subsequent PR/track in an issue)?

Reply:
This is not a change in behaviour. See here https://github.com/googleforgames/quilkin/blob/a439516e16199857bda8a2d2a8c3c1bfcf47f37a/src/filters/capture_bytes.rs#L64-L73 the packets are dropped when suffix or prefix don't have enough bytes.

Now whether this behaviour is desirable is a different story. I agree that this filter should probably only be able capturing, and make it configurable as to whether it drops on failing captures. I can add that change to this PR if you agree.

Reply:
ooooh yeah, I see your point. 👍🏻 Oh that's load of fun. Prefix and Suffix are checking for a match just by the nature of checking if there is enough data to actually grab. Excellent point, I didn't think of it that way.

> Now whether this behaviour is desirable is a different story. I agree that this filter should probably only be capturing, and make it configurable as to whether it drops on failing captures. I can add that change to this PR if you agree.

I would absolutely be happy for this addition. I'll leave it up to you if you want to do it in this PR, or we can create an issue and track it as a separate unit of work.